RavenDB-17262 Create a rvn create-setup-package command

omerlewitz opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-17262
https://github.com/ravendb/ravendb/pull/13408

### Additional description

Major changes are in the last commit.
Most of the work done is -  making sure I split the classes correctly to reduce 

1. Reducing classes size
2. Ensure that `ServerStore` arguments are changed so when creating `class links` in `rvn.csproj` I won't have to link the while server.
3. The requested changes in the last PR resolved

### Type of change

- New feature

### How risky is the change?

- Moderate 

### Backward compatibility

- Ensured. SetupManager was tested manually to make sure it didn't break.

### Is it platform-specific issue?

- No

### Documentation update

- No documentation update is needed 

### Testing 

- Tests have been added that prove the fix is effective or that the feature works


### Is there any existing behavior change of other features due to this change?


- No

### UI work

- No UI work is needed

Commit:
RavenDB-17262 - LetsEncryptUtils helper class and minor changes in CommandLineApp.cs and AdminCertificatesHandler.cs

Commit:
RavenDB-17262 - Removing redundant code resolving conversations

Commit:
RavenDB-17262 - Reducing large classes, starting to export methods to other classes in order to reduce nested class links in RVN cs.proj

Commit:
RavenDB-17262 - Removing server store parameter in order to avoid linking rvn project to redundant classes

Commit:
RavenDB-17262 - Minor changes - removing server store argument

Commit:
RavenDB-17262 - CanGetLetsEncryptCertificateAndRenewIt test failure fixed as well as re-basing.

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+
+namespace Raven.Server.Commercial.LetsEncrypt
+{
+    public static class LetsEncryptByTools
```
`ByTools` ?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+            progress?.AddInfo("Setting up RavenDB in Let's Encrypt security mode.");
+
+            if (ZipFileHelper.IsValidEmail(setupInfo.Email) == false)
+                throw new ArgumentException("Invalid e-mail format" + setupInfo.Email);
```
```suggestion
                throw new ArgumentException("Invalid e-mail format: " + setupInfo.Email);
```

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+            }
+            catch (Exception e)
+            {
+                throw new InvalidOperationException("Failed to initialize lets encrypt challenge: " + e);
```
```suggestion
                throw new InvalidOperationException("Failed to initialize lets encrypt challenge", e);
```

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+                };
+                var serializeObject = JsonConvert.SerializeObject(registrationInfo);
+
+                foreach (var action in DnsBridgeActions)
```
Why are we calling them in a loop? You are hitting 3 different endpoints with the same request, what are you doing here?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+        private const string AcmeClientUrl = "https://acme-v02.api.letsencrypt.org/directory";
+        private static readonly string[] DnsBridgeActions = {"user-domains", "domain-availability", "claim"};
+
+        public static async Task<byte[]> SetupLetsEncryptByRvn(SetupInfo setupInfo, string settingsPath, SetupProgressAndResult progress, string dataFolder, CancellationToken token)
```
`ByRvn`? What does that mean?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptByTools.cs:
```diff
@@ -0,0 +1,198 @@
+            }
+        }
+
+        public static async Task<byte[]> SetupOwnCertByRvn(SetupInfo setupInfo, string settingsPath, SetupProgressAndResult progress, CancellationToken token)
```
`ByRvn`?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+            case SetupMode.LetsEncrypt when ZipFileHelper.IsValidEmail(setupInfo.Email) == false:
+                throw new ArgumentException("Invalid email address.");
+            case SetupMode.LetsEncrypt when IsValidDomain(setupInfo.Domain + "." + setupInfo.RootDomain) == false:
+                throw new ArgumentException("Invalid domain name.");
```
What is the domain name?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+            case SetupMode.LetsEncrypt when setupInfo.NodeSetupInfos.ContainsKey(setupInfo.LocalNodeTag) == false:
+                throw new ArgumentException($"At least one of the nodes must have the node tag '{setupInfo.LocalNodeTag}'.");
+            case SetupMode.LetsEncrypt when ZipFileHelper.IsValidEmail(setupInfo.Email) == false:
+                throw new ArgumentException("Invalid email address.");
```
What is the email?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+        switch (setupMode)
+        {
+            case SetupMode.LetsEncrypt when setupInfo.NodeSetupInfos.ContainsKey(setupInfo.LocalNodeTag) == false:
+                throw new ArgumentException($"At least one of the nodes must have the node tag '{setupInfo.LocalNodeTag}'.");
```
What values do we have?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+            case SetupMode.LetsEncrypt when IsValidDomain(setupInfo.Domain + "." + setupInfo.RootDomain) == false:
+                throw new ArgumentException("Invalid domain name.");
+            case SetupMode.LetsEncrypt when setupInfo.ClientCertNotAfter.HasValue && setupInfo.ClientCertNotAfter <= DateTime.UtcNow.Date:
+                throw new ArgumentException("The client certificate expiration date must be in the future.");
```
What is the value?

ayende commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        }
+        catch (Exception e)
+        {
+            throw new InvalidOperationException("Failed to create setting file(s).", e);
```
```suggestion
            throw new InvalidOperationException("Failed to create settings file(s).", e);
```

Commit:
RavenDB-17262 - Requested changes adjustments

Commit:
RavenDB-17262 - Removing redundant code

gregolsky commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+﻿using System;
```
I'd call the file `SettingsZipFileHelper`. What do you think?

omerlewitz replied:
I agree.

arekpalinski replied:
Agree. `ZipFileHelper` is too generic

gregolsky commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+    
+    public static string IndentJsonString(string json)
```
Since it has to do with Json and not Zip directly and also is used in other places (LicenseManager) can we make JsonStringHelper and move that there?

omerlewitz replied:
done

gregolsky commented on src/Raven.Server/Commercial/SetupInfo.cs:
```diff
@@ -342,7 +343,7 @@ public void AddError(string message, Exception ex = null)
             Message = $"[{SystemTime.UtcNow:T} {type}] {message}";
             Messages.Enqueue(Message);
-            if (Logger.IsInfoEnabled)
+            if (Logger is {IsInfoEnabled: true})
```
I'd prefer `Logger?.IsInfoEnabled`.

omerlewitz replied:
the implementation has been changed

gregolsky commented on src/Raven.Server/ServerWide/SecretProtection.cs:
```diff
@@ -324,9 +326,107 @@ private static byte[] DecryptProtectedData(byte[] secret, byte[] key)
+            }
+        }
+
+        public static CertificateUtils.CertificateHolder ValidateCertificateAndCreateCertificateHolder(string source, X509Certificate2 loadedCertificate, byte[] rawBytes, string password, LicenseType licenseType, bool certificateValidationKeyUsages)
```
I'd rename `certificateValidationKeyUsages` parameter to `validateCertKeyUsages`.

gregolsky replied:
Previously value for `certificateValidationKeyUsages` was taken from `serverStore.Configuration.Security.CertificateValidationKeyUsages`.

Is there a behavior change?

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,92 @@ public static int Run(string[] args)
+                                ExitWithError("--path option must be set",cmd);
+                                break;
+                            case ImportCertificateMode:
+                                return await SetupPackageConfiguratorLetsEncrypt(setupParamVal, packageOutParamVal, cmd, ImportCertificateMode, pathToCertVal, certPassVal, token);
```
I guess since that's ImportCertificateMode it doesn't have anything to do with LE. Maybe we should rename SetupPackageConfiguratorLetsEncrypt to something more general e.g. CreateSetupPackage?

omerlewitz replied:
I agree

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,92 @@ public static int Run(string[] args)
+                var pathToCert = ConfigureCertPath(cmd);
+                var certPass = ConfigureCertPassword(cmd);
+             
+                    cmd.OnExecuteAsync(async token =>
```
wrong code indent?

omerlewitz replied:
yeap, the structure changed

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,92 @@ public static int Run(string[] args)
+                    if (packageOutParam == null)
+                        packageOutParam = setupInfo.Domain + "-package.zip";
+                    else
+                        packageOutParam += "-package.zip";
```
If the user specifies an output path with `packageOutParam` I don't think we need to do anything else with that and just output it where we're told to?

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,92 @@ public static int Run(string[] args)
+                    else
+                        packageOutParam += "-package.zip";
+
+                    if (mode is ImportCertificateMode)
```
We should probably use switch on the `mode` here and default to throw NotSupportedException. The decision on the default value of `mode` should be made in the CLI option configuration itself.

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -342,7 +441,30 @@ private static int ExitWithError(string errMsg, CommandLineApplication cmd)
 
+        private static CommandOption ConfigurePackageOutputFile(CommandLineApplication cmd)
+        {
+            return cmd.Option("-o|--package-output-file", "default output file", CommandOptionType.SingleValue);
```
```suggestion
            return cmd.Option("-o|--package-output-path", "setup package output path", CommandOptionType.SingleValue);
```

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,92 @@ public static int Run(string[] args)
+                        zipFile = await LetsEncryptRvnUtils.SetupLetsEncrypt(setupInfo, setupParamsPath, progress, string.Empty ,token);
+                    }
+                }
+                var path = Path.Combine(AppContext.BaseDirectory, packageOutParam);
```
Instead of changing using app's base dir, we should use what's passed in `packageOutParam` alone. User would in general expect that if he run this in his home dir with
`/opt/ravendb/RavenDB/Server/rvn -o /tmp/setup_package.zip`

The resulting package would be in `/tmp/setup_package.zip`.
The app should not care where we're at the moment. Also if a relative path is specified e.g. `./package.zip` then just save it there.

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -342,7 +441,30 @@ private static int ExitWithError(string errMsg, CommandLineApplication cmd)
+        
+        private static CommandOption ConfigureCertPassword(CommandLineApplication cmd)
+        {
+            return cmd.Option("-p|--password", "Certificate password", CommandOptionType.SingleValue);
```
Passing passwords as CLI arguments is not considered safe usually - it gets into commands history, is visible on `tty`s etc. Would be nice to allow to provide an alternative way - add ability to pass it through an enviroment variable.

omerlewitz replied:
tnx. I also removed the `-p`. It may confuse the user with `-p` as if it is the path to the `certificate`

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptRvnUtils.cs:
```diff
@@ -0,0 +1,194 @@
+
+namespace Raven.Server.Commercial.LetsEncrypt
+{
+    public static class LetsEncryptRvnUtils
```
Why include `Rvn` here? It's not rvn tool specific helper - we do use it during the regular setup with the usage of Let's Encrypt, don't we?

omerlewitz replied:
agree

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+                setupInfo.NodeSetupInfos[key].TcpPort = 38888;
+
+            if (setupMode == SetupMode.LetsEncrypt &&
+                setupInfo.NodeSetupInfos[key].Addresses.Any(ip => ip.Equals("0.0.0.0")) &&
```
We have `"0.0.0.0"` hard coded in multiple places here - maybe introduce a const for that? This way it's easier and faster to find all usages of that

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptValidationApiHelper.cs:
```diff
@@ -0,0 +1,183 @@
+        {
+            case SetupMode.LetsEncrypt when setupInfo.NodeSetupInfos.ContainsKey(setupInfo.LocalNodeTag) == false:
+                throw new ArgumentException($"At least one of the nodes must have the node tag '{setupInfo.LocalNodeTag}'. Nodes: " + setupInfo.NodeSetupInfos.Keys);
+            case SetupMode.LetsEncrypt when ZipFileHelper.IsValidEmail(setupInfo.Email) == false:
```
`ZipFileHelper` validates the email? I doesn't sound like correct responsibility.

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/RavenDnsRecordHelper.cs:
```diff
@@ -0,0 +1,316 @@
+{
+        private const string GoogleDnsApi = "https://dns.google.com";
+
+        public class UpdateDnsRecordParameters
```
Wouldn't make sense to make it standalone class? It's public, used in few places outside and right now you need to access it via `RavenDnsRecordHelper.`

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        return str;
+    }
+    
+    public class CompleteAuthorizationAndGetCertificateParameters
```
Similar issue here. The instance of this class is created in a few places - seems to make sense to move it to a dedicated `CompleteAuthorizationAndGetCertificateParameters.cs` so you won't need to access it via helper.

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        public RSA ExistingPrivateKey;
+    }
+    
+    public static bool IsValidEmail(string email)
```
It's not responsibility of that helper. You can move where it would suit better (maybe to `LetsEncryptUtils` ?) or create a dedicated static class like e.g. `SetupValidator` or `LetsEncryptSetupValidator` - not sure if this is validation specific to let's encrypt setup mode or not

omerlewitz replied:
changed

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        }
+    }
+
+    public static string IpAddressToUrl(string address, int port)
```
Is is used outside this helper? Can it be private method?

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/ZipFileHelper.cs:
```diff
@@ -0,0 +1,628 @@
+        return url;
+    }
+
+    public  static string IpAddressToTcpUrl(string address, int port)
```
Can it be private?

omerlewitz replied:
done

arekpalinski reviewed (commented):
The structure of classes and referencing them via links looks much better now. That's great.

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptCertificateUtil.cs:
```diff
@@ -0,0 +1,178 @@
+
+public class LetsEncryptCertificateUtil
+{
+    public static string GetServerUrlFromCertificate(X509Certificate2 cert, SetupInfo setupInfo, string nodeTag, int port, int tcpPort, out string publicTcpUrl, out string domain)
```
Is there anything specific to Let's Encrypt cert here? If not and it's more generic then maybe it could be part of `CertificateUtils`?

omerlewitz replied:
The main usage is within the `zipFileHelper`, Yes, it's better to move it to a more generic place.

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/SettingsZipFileHelper.cs:
```diff
@@ -0,0 +1,582 @@
+
+public static class SettingsZipFileHelper
+{
+    public static async Task<byte[]> CompleteClusterConfigurationAndGetSettingsZip(CompleteClusterConfigurationParameters parameters)
```
I wonder if the part "CompleteClusterConfiguration" should be handled by `SettingsZipFileHelper`? IMO that helper should just be responsible for "GetSettingsZip" part. 

Are we able to separate that?

omerlewitz replied:
Lets consider this:
- Moving `CompleteAuthorizationAndGetCertificate` to `LetsEncryptUtils` it is more suitable.
- Moving `BuildPfx` to `CertificateUtils`
- Changing `WriteSettingsJsonLocally` to `WriteJsonFile` and will be under `LocalWrites` or `WriteToDisk`
I'm not sure myself what's better than keeping it in the way it was...

omerlewitz replied:
I made changes to the function.
I think it's much more readable and "cleaner"

arekpalinski commented on src/Raven.Server/Commercial/EmailValidator.cs:
```diff
@@ -0,0 +1,21 @@
+
+public class EmailValidator
+{
+    public static bool IsValidEmail(string email)
```
Minor - It's already in `EmailValidator` so just `IsValid(string email)` should be enough

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptCertificateUtil.cs:
```diff
@@ -0,0 +1,178 @@
+    }
+    
+    public static async Task WriteCertificateAsPemAsync(string name, byte[] rawBytes, string exportPassword, ZipArchive archive)
+        {
```
Formatting issue:


```
public static async Tas
        {
```

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/JsonStringHelper.cs:
```diff
@@ -0,0 +1,19 @@
+
+public class JsonStringHelper
+{
+    public static string IndentJsonString(string json)
```
Just Indent? So instead of `JsonStringHelper.IndentJsonString()` you'd have `JsonStringHelper.Indent()` - it's easier to read it.

omerlewitz replied:
Ok

arekpalinski commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -15,6 +21,9 @@ internal static class CommandLineApp
         private static CommandLineApplication _app;
 
+        private const string LetsEncryptMode = "lets-encrypt";
+        private const string ImportCertificateMode = "import-certificate";
```
So we'll have:

`rvn create-setup-package lets-encrypt`
`rvn create-setup-package import-certificate {path-to-cert}` 


Is that correct? 

I have some doubts if `import-certificate` is the correct name for that setup mode. Other options: `with-certificase`, `self-signed-certificate` or just `secured` like the enum value of `SetupMode`?

Thoughts? CC @gregolsky

arekpalinski replied:
And just to make sure - those options will be printed to the user when --help will be used, right?

gregolsky replied:
@arekpalinski From what I see it's gonna be:

```
rvn create-setup-package --mode lets-encrypt --setup-parameters-path setup.json
rvn create-setup-package -m lets-encrypt -s setup.json
rvn create-setup-package import-certificate --cert-path cert.pfx
```

gregolsky replied:
Maybe `own-certificate`?

And yes, `--help` is generated autiomatically.

arekpalinski replied:
I think `own-certificate` suits better than `import-certificate`

omerlewitz replied:
@arekpalinski  regarding the `enum`.
Creating an `enum` will cause me to add `[Description]` property which will cause trying to parse the string using reflection for just a string I'm getting from the user. 
I couldn't make it work. I think the better choice, in this case, is just using two consts.

Am I missing something here ?

arekpalinski replied:
okay, if that causes too much hassle then let's keep it simple with the usage of const

arekpalinski commented on tools/rvn/Parameters/CreateSetupPackageParameters.cs:
```diff
@@ -0,0 +1,17 @@
+    public string SetupInfoPath;
+    public string PackageOutputPath;
+    public CommandLineApplication Command;
+    public string SetupMode;
```
Why is that a string instead of `Raven.Server.Commercial.SetupMode` enum?

gregolsky replied:
@arekpalinski it's different setup mode
https://github.com/ravendb/ravendb/pull/13563/files/cce45d0e060fdccdc0ba288c722327179a5f82e2#diff-ad9cbee56ef20d0464113960dae85d9641aae42f0fa937b47e2e599ea421fcdeR71

arekpalinski replied:
Okay. Still wouldn't be easier to deal with enum instead of string?

omerlewitz replied:
@arekpalinski I agree...

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+                    var setupParamVal = setupParam.Value();
+                    var packageOutPathVal = packageOutPath.Value();
+                    var certPathVal = certPath.Value();
+                    var certPassTuple = (certPass.Value(), Environment.GetEnvironmentVariable("RVN_CERT_PASS"));
```
how about doing
```
certPass.Value() ?? Environment.GetEnvironmentVariable("RVN_CERT_PASS")
```

then you dont need a tuple?

omerlewitz replied:
tnx

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+            }
+            catch (Exception e)
+            {
+                return ExitWithError($"Failed to deserialize SetupInfo object form this file path {nameof(parameters.SetupInfoPath)}\nError: {e}",parameters.Command);
```
```suggestion
                return ExitWithError($"Failed to deserialize SetupInfo object form this file path {nameof(parameters.SetupInfoPath)}\nError: {e}", parameters.Command);
```

omerlewitz replied:
tnx

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+            try
+            {
+                if (Path.GetExtension(parameters.PackageOutputPath)?.Equals(".zip") == false)
+                    return ExitWithError($"'{parameters.PackageOutputPath}' file name must end with ZIP extenstion", parameters.Command);
```
User would probably like to know that before the whole behemoth getting the lets encrypt certificate, contacting api and generating settings kicks off :) As a user I would be pissed if I waited few minutes just to get an error message I passed a wrong filename. Also make sure it's case-insensitive.

omerlewitz replied:
you are right

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+            byte[] zipFile;
+            SetupInfo setupInfo;
+
+            if (File.Exists(parameters.SetupInfoPath) == false)
```
Let's move all the validation of parameters to earliest executed piece possible.

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+                }
+                case ImportCertificateMode:
+                {
+                    
```
```suggestion          
```

omerlewitz replied:
fixed

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,137 @@ public static int Run(string[] args)
+                    var certBytes = await File.ReadAllBytesAsync(parameters.CertificatePath, parameters.CancellationToken);
+                    var certBase64 = Convert.ToBase64String(certBytes);
+                    setupInfo.Certificate = certBase64;
+                    zipFile = await LetsEncryptUtils.ImportCertificateSetup(setupInfo, parameters.Progress, parameters.CancellationToken);
```
'Bring your own certificate' option doesn't have anything to do with LetsEncrypt, does it? Maybe it should reside in another class or the class should have a different name.

omerlewitz replied:
done

gregolsky commented on src/Raven.Server/Commercial/LetsEncrypt/SettingsZipFileHelper.cs:
```diff
@@ -0,0 +1,582 @@
+                    else
+                        certPath = Path.Combine(AppContext.BaseDirectory, certificateFileName);
+                    
+                    if (parameters.SetupInfo.ModifyLocalServer)
```
This should be `false` when used from `rvn create-setup-package`. I see that `SetupInfo` is just parsed from file, so I guess that's the default. 

Could you add an assertion to tests that makes sure settings.json is *not* modified when running from `rvn` and is modified when run normally during the studio setup?

omerlewitz replied:
done

gregolsky commented:
@garayx I believe that once it's merged this will require regression testing - doing manual setup (all 3 options) - but also trying out new rvn command. Let's talk it through on the leaders meeting.

cc @ppekrol 

EDIT: Talked with Marcin and Danielle. She will conduct those tests once the PR is merged.

Commit:
RavenDB - 17262 - Changes listed below

* CompleteClusterConfiguration function - soc
* CommandLineApp - Minor changes
* CompleteAuthorizationAndGetCertificate - Changed unnecessary Task returned types
* LetsEncryptSetupUtils, ImportCertificateSetupUtils and CompleteClusterConfiguration - In more suitable  place

gregolsky commented on src/Raven.Server/Commercial/SetupManager.cs:
```diff
@@ -1,86 +1,61 @@
 using Raven.Server.Utils;
-using Raven.Server.Utils.Cli;
-using Raven.Server.Web.Authentication;
+using rvn.Server.SetupWizard;
```
I don't think the server should reference back rvn namespaces.

omerlewitz replied:
fixed

gregolsky commented on src/Raven.Server/Config/RavenConfiguration.cs:
```diff
@@ -45,6 +45,8 @@ public class RavenConfiguration
 
         public CoreConfiguration Core { get; }
 
+        public SecurityConfiguration Security { get; }
```
What does this one do? This is just to get `Security` options into rvn right?

cc @ppekrol

ppekrol replied:
correct

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/CompleteAuthorizationAndGetCertificateParameters.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Security.Cryptography;
+using System.Threading;
+
+namespace Raven.Server.Commercial.LetsEncrypt;
```
Is it correct namespece? Is this class specific to Let's encrypt? If not then maybe `Raven.Server.Commercial.SetupWizadr` would be better location for it

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/JsonStringHelper.cs:
```diff
@@ -0,0 +1,19 @@
+﻿using System.IO;
+using Newtonsoft.Json;
+
+namespace Raven.Server.Commercial.LetsEncrypt;
```
Please move it to `Raven.Server/Utils` dir and adjust the namespace

gregolsky commented on src/Raven.Server/RavenServer.cs:
```diff
@@ -1341,7 +1346,7 @@ private async Task<byte[]> RenewLetsEncryptCertificate(CertificateHolder existin
 
             SecretProtection.ValidateCertificateAndCreateCertificateHolder("Let's Encrypt Refresh", cert, certBytes,
-                setupInfo.Password, ServerStore);
+                setupInfo.Password, ServerStore.LicenseManager.LicenseStatus.Type, true);
```
I noticed `ServerStore.LicenseManager.LicenseStatus.Type` is used 6 times in this PR. Maybe instead of using dots everytime we could add a getter or make an extension method for this e.g. `ServerStore.GetLicenseType()`

omerlewitz replied:
okay

arekpalinski commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptSimulationHelper.cs:
```diff
@@ -0,0 +1,178 @@
+
+                        if (addresses.Length == 0)
+                        {
+                            var defaultIp = new IPEndPoint(IPAddress.Parse("0.0.0.0"), port == 0 ? 443 : port);
```
Don't we already have const for "0.0.0.0"? We could use it here

omerlewitz replied:
Now it's within `Client.Constants`

gregolsky commented on src/Raven.Server/ServerWide/SecretProtection.cs:
```diff
@@ -324,9 +326,116 @@ private static byte[] DecryptProtectedData(byte[] secret, byte[] key)
+        {
+            if (loadedCertificate.NotAfter < DateTime.UtcNow)
+            {
+                string msg = $"The provided certificate {loadedCertificate.FriendlyName} from {source} is expired! Thumbprint: {loadedCertificate.Thumbprint}, Expired on: {loadedCertificate.NotAfter}";
```
We could add quotes around the name.
```suggestion
                string msg = $"The provided certificate '{loadedCertificate.FriendlyName}' from {source} is expired! Thumbprint: {loadedCertificate.Thumbprint}, Expired on: {loadedCertificate.NotAfter}";
```

omerlewitz replied:
done

gregolsky commented on src/Raven.Server/ServerWide/SecretProtection.cs:
```diff
@@ -324,9 +326,116 @@ private static byte[] DecryptProtectedData(byte[] secret, byte[] key)
+                if (loadedCertificate.NotAfter > DateTime.UtcNow.AddMonths(4))
+                {
+                    const string msg = "The server certificate expiration date is more than 4 months from now. " +
+                                       "This is not allowed when using the developer license. " +
+                                       "The developer license is not allowed for production use. " +
```
```suggestion
                                       "This is not allowed when using Developer license. " +
                                       "Developer license is not allowed for production use. " +
```

omerlewitz replied:
done

arekpalinski commented on src/Raven.Server/Commercial/SetupWizard/ImportCertificateSetupUtils.cs:
```diff
@@ -0,0 +1,61 @@
+
+namespace Raven.Server.Commercial.SetupWizard;
+
+public class ImportCertificateSetupUtils
```
I wonder if `OwnCertificateSetupUtils` wouldn't be more accurate. This mode is more about providing own certificate while "import" might suggest it's about importing the certificate to already setup cluster. Also on UI we use "Provide your own certificate" (nothing about import).

omerlewitz replied:
Changed

arekpalinski reviewed (commented):
Looks very nice. Everything is logically organized and it's easier to understand the setup flow when reading the code. 

I left a few minor comments. Please note there is one conflict.

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +59,130 @@ public static int Run(string[] args)
+            
+            if (Path.GetExtension(parameters.PackageOutputPath)?.Equals(".zip", StringComparison.OrdinalIgnoreCase) == false)
+            {
+                return ExitWithError($"'{parameters.PackageOutputPath}' file name must end with ZIP extenstion", parameters.Command);
```
```suggestion
                return ExitWithError($"'{parameters.PackageOutputPath}' file name must end with ZIP extension", parameters.Command);
```

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +59,130 @@ public static int Run(string[] args)
+            }
+            catch (Exception e)
+            {
+                return ExitWithError($"Failed to deserialize SetupInfo object form this file path {nameof(parameters.SetupInfoPath)}\nError: {e}", parameters.Command);            }
```
```suggestion
                return ExitWithError($"Failed to load RavenDB Setup parameters from file: {parameters.SetupInfoPath}{Environment.NewLine}Error: {e}", parameters.Command);            }
```

gregolsky replied:
Also we need to write in this command's `--help` how does a setup parameters file look like.

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +59,130 @@ public static int Run(string[] args)
+
+            switch (parameters.Mode)
+            {
+                case "own-certificate" when string.IsNullOrEmpty(parameters.CertificatePath):
```
Why don't we use constants for these options in `case`?

omerlewitz replied:
tnx

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -338,11 +473,36 @@ private static void ConfigureOfflineOperationCommand()
         private static int ExitWithError(string errMsg, CommandLineApplication cmd)
         {
-            cmd.Error.WriteLine(errMsg);
+            cmd.Out.WriteLine();
```
If we're writing an error we should probably write new lines to stderr, not stdout.

omerlewitz replied:
done

gregolsky replied:
did you push it? It's still written to Out instead of Error here

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -338,11 +473,36 @@ private static void ConfigureOfflineOperationCommand()
+        
+        private static CommandOption ConfigureSetupParameters(CommandLineApplication cmd)
+        {
+            return cmd.Option("-s|--setup-parameters-path", "Call setup endpoints and obtain the setup package for setup up the cluster", CommandOptionType.SingleValue);
```
We need to describe what the file contents need to look like.

gregolsky commented on src/Raven.Server/Utils/IOExtensions.cs:
```diff
@@ -130,6 +130,7 @@ public static void DeleteDirectory(string directory)
                     }
                     catch (IOException)
                     {
+                        Console.WriteLine("Got to delete");
```
?

omerlewitz replied:
Removed

Commit:
RavenDB-17262 - Listed changes below:

1) *_parameters classes are stand alone.
2) CommandLineApp - more readable design and adding error messages
3) SetupProgressAndResult - The change due to the fact that in RVN we don't need logger.
4) SecretProtection - Adding messages for RVN usages

gregolsky commented on src/Raven.Client/Constants.cs:
```diff
@@ -119,6 +119,16 @@ private Certificates()
+            public const int DefaultSecuredRavenDbHttpPort = 443;
+            public const int DefaultUnsecuredRavenDbHttpPort = 8080;
+            public const int DefaultSecuredRavenDbTcpPort = 38888;
+            public const int DefaultUnsecuredRavenDbTcpPort = 38881;
```
I believe that unsecured tcp port is the same - 38888.

omerlewitz replied:
I will check it again.
Regarding the `--help` of `-s` option, I think we will leave it for the documentation.
I did some research given the link you sent me and in any option, the JSON doesn't look good.
We can place it after the help text chuck but it does not seem to be the right thing to do.
Let's leave it for the documentation for now

gregolsky replied:
OK. What about putting it in an `rvn.md` text file next to it?

omerlewitz replied:
@gregolsky  I removed it for now

arekpalinski commented on tools/rvn/rvn.csproj:
```diff
@@ -19,7 +19,22 @@
+    <Compile Include="..\..\src\Raven.Client\Documents\Operations\OperationState.cs" Link="Client\Documents\Operations\OperationState.cs" />
+    <Compile Include="..\..\src\Raven.Client\Documents\Operations\IOperationProgress.cs" Link="Client\Documents\Operations\IOperationProgress.cs" />
+    <Compile Include="..\..\src\Raven.Client\Util\SystemTime.cs" Link="Client\Util\SystemTime.cs" />
+    <Compile Include="..\..\src\Raven.Client\Util\SystemTime.cs" Link="Client\Util\SystemTime.cs" />
```
Here you have duplicated line which causes compilation error:

https://github.com/ravendb/ravendb/runs/5124495174?check_suite_focus=true#step:7:61

omerlewitz replied:
Weird.I didn't get the error when I compiled the project

Commit:
RavenDB-17262 - removing duplication rvn/rvn.csproj

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -51,6 +61,130 @@ public static int Run(string[] args)
+
+            switch (parameters.Mode)
+            {
+                case OwnCertificate when string.IsNullOrEmpty(parameters.CertificatePath):
```
One last thing. 
Can we move validation pieces before this mode switch? We should have 3 clear cases:
OwnCertificate (no whens), 
LetsEncrypt (no when), 
default (ExitWithError("Invalid mode provided").

Commit:
RavenDB-17262 - Moving setup mode validations to a function

arekpalinski commented:
@omerlewitz please note there is a conflict on one file

arekpalinski commented:
@omerlewitz please note there one conflict

arekpalinski commented:
@omerlewitz please also rebase the PR (we have updated the .net version and the tests have failed - they will run correctly after the rebase)

Commit:
RavenDB-17262 - Minor adjustments. SecretProtection - avoid using ServerStore argument(rvn.csproj)

gregolsky commented:
Please resolve conflicts.
Also it stopped to compile - see https://github.com/ravendb/ravendb/runs/5244978413?check_suite_focus=true

Commit:
RavenDB-17262 - Redundant code removed, minor adjustments, naming convention.

Commit:
RavenDB - 17262 - last changes

Commit:
RavenDB - 17262 - Name-spaces fixed

Commit:
RavenDB-17262 - Fixing build and re-basing

Commit:
RavenDB-17262 - Requested changes regarding Danielle's review

Commit:
RavenDB-17262 - Requested changes regarding Danielle's second review

Commit:
RavenDB-17262 - re-base and applying stash

ppekrol commented:
Does not compile, please fix that.

Commit:
RavenDB-17262 - Greg's Linux review - RavenDB-18361 & RavenDB-18362 as well as fixing compilation error due to re-basing

Commit:
RavenDB-17262 - minor adjustments while testing the branch in Linux

gregolsky commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptCertificateUtil.cs:
```diff
@@ -0,0 +1,101 @@
+
+public class LetsEncryptCertificateUtil
+{
+    internal static (byte[] CertBytes, CertificateDefinition CertificateDefinition, X509Certificate2 SelfSignedCertificate) GenerateCertificateTask(CertificateUtils.CertificateHolder certificateHolder, string certificateName, SetupInfo setupInfo)
```
I think we should indicate we're generating *client* certificate here?

omerlewitz replied:
fixed

gregolsky commented on src/Raven.Server/Commercial/LetsEncrypt/LetsEncryptCertificateUtil.cs:
```diff
@@ -0,0 +1,101 @@
+        return (certBytes, newCertDef, selfSignedCertificate);
+    }
+
+    public static async Task WriteCertificateAsPemAsync(string name, byte[] rawBytes, string exportPassword, ZipArchive archive)
```
This once kinda mixes certificate conversion and writing to zip file. At least let's maybe name it `WriteCertificateAsPemToZipArchiveAsync`?

omerlewitz replied:
done

gregolsky commented on src/Raven.Server/Commercial/SetupManager.cs:
```diff
@@ -266,960 +275,257 @@ public static async Task<IOperationResult> ContinueClusterSetupTask(Action<IOper
+        })
+        {
+            Processed = 0,
+            Total = 4
```
Out of curiosity, what's this number?

gregolsky commented on src/Raven.Server/ServerWide/SecretProtection.cs:
```diff
@@ -325,9 +328,117 @@ private static byte[] DecryptProtectedData(byte[] secret, byte[] key)
+                string msg = $"The provided certificate '{loadedCertificate.FriendlyName}' from {source} is expired! Thumbprint: {loadedCertificate.Thumbprint}, Expired on: {loadedCertificate.NotAfter}";
+                if (Logger.IsOperationsEnabled)
+                    Logger.Operations(msg);
+                progress?.AddError(msg);
```
We log but we don't throw? The original code also had this issue
![image](https://user-images.githubusercontent.com/1541805/160126477-2808b846-221c-444f-a308-0a2f7150137e.png)

omerlewitz replied:
thanks
fixed

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -52,6 +63,209 @@ public static int Run(string[] args)
+            using (StreamReader file = File.OpenText(parameters.SetupJsonPath))
+            {
+                JsonSerializer serializer = new();
+                setupInfo = (SetupInfo)serializer.Deserialize(file, typeof(SetupInfo));
```
We try/catch and throw in case there's JSON deserialization error.

omerlewitz replied:
I thought I handled that already.
Thanks for noticing.

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -52,6 +63,209 @@ public static int Run(string[] args)
+
+            foreach (var tag in setupInfo.NodeSetupInfos.Keys.Where(tag => IsValidNodeTag(tag) == false))
+            {
+                throw new InvalidOperationException($"{tag} - node tag must contain only capital letters. Maximum length should be up to 4 characters");
```
Shouldn't we pass all tags that are not valid in this exception?

omerlewitz replied:
done

gregolsky commented on tools/rvn/CommandLineApp.cs:
```diff
@@ -52,6 +63,209 @@ public static int Run(string[] args)
+
+                if (nodeInfoNode.Port == 0)
+                {
+                    nodeInfoNode.Port = Raven.Client.Constants.Network.DefaultSecuredRavenDbHttpPort;
```
Since this method has side-effects like this we could rename it to "ValidateSetupInfoAndSetDefaultSetupParametersIfNeeded"?

omerlewitz replied:
done

Commit:
RavenDB-17262 - Refactoring and sorting error messages

omerlewitz commented:
@ppekrol  will resolve it today

Commit:
RavenDB-17262: started working on stub implementation

ppekrol commented on src/Raven.Client/Constants.cs:
```diff
@@ -119,6 +119,15 @@ private Certificates()
             public const int MaxNumberOfCertsWithSameHash = 5;
         }
+        
+        public class Network
```
let's make it internal for now

omerlewitz replied:
fixed

ppekrol commented on test/Tests.Infrastructure/TestBase.cs:
```diff
@@ -182,6 +186,122 @@ public void DoNotReuseServer(IDictionary<string, string> customSettings = null)
 
+        protected static TestCertificatesHolder _selfSignedCertificates;
+
+        protected TestCertificatesHolder GenerateAndSaveSelfSignedCertificate(bool createNew = false, [CallerMemberName] string caller = null)
```
This is a copy of a method from RavenTestBase.Certificates. You should use it or move it to this class, depending why you need it here.

omerlewitz replied:
fixed

Commit:
RavenDB-17262 - Pawel's remarks

Commit:
RavenDB-17262 - Major changes - adding support to claim DNS before starting lets encrypt task.

Adding only necessary classes links cs.proj

Changing functions scope from !RVN macro

Raven/Utils/ - Server store argument removed.

Adding tests

Operations/Operations.cs - redundant code I added long time ago.

LetsEncrypt/LetsEncryptSimulationHelper.cs - Select func is obsolete

LetsEncrypt/RavenDnsRecordHelper.cs - null checking

ZipFileHelper - redundant code removed

SetupManager - restoring extactZipFile function

Program,RavenCLI,WelcomeMessage - removing server store usage

Authentication/AuthenticationLetsEncryptTests.cs - removed redundant using

SlowTests.csproj - Removing unnecessary linking.

Raven.Server.csproj - Removing unnecessary linking
